planner/core/logical_plan_builder.go

```diff
@@ -3380,17 +3381,27 @@ func (b *PlanBuilder) handleDefaultFrame(spec *ast.WindowSpec, windowFuncName st
 				End:   ast.FrameBound{Type: ast.CurrentRow},
 			},
 		}
-		return &newSpec, true
+		return &newSpec, true, nil
 	}
 	// For functions that operate on the entire partition, the frame clause will be ignored.
 	if !needFrame && spec.Frame != nil {
+		if spec.OrderBy != nil && len(spec.OrderBy.Items) > 1 && spec.Frame.Type == ast.Ranges {
+			start, end := spec.Frame.Extent.Start, spec.Frame.Extent.End
+			if start.Type == ast.Preceding && !start.UnBounded {
+				return spec, false, ErrWindowRangeFrameOrderType.GenWithStackByArgs(getWindowName(spec.Name.O))
+			}
+			if end.Type == ast.Following && !end.UnBounded {
+				return spec, false, ErrWindowRangeFrameOrderType.GenWithStackByArgs(getWindowName(spec.Name.O))
+			}
+		}
```

Comment:
Your code is for frame validation, is that suitable to be put in `handleDefaultFrame`?

Reply:
Besides, the ErrWindowRangeFrameOrderType error is `needFrame` irrelevant, considering this case
```mysql
SELECT FIRST_VALUE(fieldA) OVER w1 AS 'dense_rank', fieldA, fieldB  FROM ( SELECT `col_time` AS fieldA, `col_time` AS fieldB FROM `table20_int_autoinc` ) as t     WINDOW w1 AS ( PARTITION BY fieldB  ORDER BY fieldB DESC, fieldA DESC  RANGE BETWEEN CURRENT ROW AND 1 FOLLOWING);
```

Reply:
Actually, the `handleDefaultFrame` is a function to remove the invalidated frames. So it's reasonable to put it here.

Reply:
And there are much more frame validations in `buildWindowFunctionFrame`.

Reply:
The reasonable order for building a window function is
1) build from ast
2) validate
3) handleDefaultFrame

May also fix #11001 #11002, and make #11011 easy to fix.

Reply:
Now, the frame validations are after `groupWindowFuncs`. We have to `handleDefaultFrame` in `groupWindowFuncs`. It hard to move this validation to `buildWindowFunctionFrame`...